PostListPage.tsx 구현

Rengod95 opened the pull request:
작업 요약

- /board/post/index.tsx 경로의 PostListPage.tsx 를 구현
- Button.tsx 컴포넌트 에러 수정
- Styled Components Naming Convention  적용
- implement Data Fetching hook 

주석

네이밍 컨벤션 작업때문에 File Changed가 크게 잡힙니다. 실질적으로 확인하셔야 할 부분은 
1. @/components/Post 부분과
2. /board/post/index.tsx 부분입니다

커밋 디테일
1. https://github.com/Rengod95/Dynamic_Roadmap/pull/8/commits/aaacce59aba864662e89989440821621d2ebf81e
2. https://github.com/Rengod95/Dynamic_Roadmap/pull/8/commits/d14cb0a9aa591db8040f89195c09ed538d7dddbb

PR 반영 
https://github.com/Rengod95/Dynamic_Roadmap/pull/8/commits/2c0e6a0a85a145d15f0a32633a2a6d323365ac68

Rengod95 committed:
feat: PostListPage 구현
- Button.tsx 수정
  - variant 와 size 를 통해 스타일링이 적용될 때 handler 함수가 동작하지 않는 부분 수정
  - fontSize의 타입 수정
  - size 입력값을 항상 px로 받도록 변경

- Icon.tsx 수정
  - path element의 인자로 viewBoxSize 속성이 들어가던 부분을 수정

- usePosts 훅 작성
  - useQuery를 이용해 데이터 패칭을 할 수 있도록 로직 분리
  - Fetched data 의 total값을 기준으로 최대 페이지 수를 관리
  - state를 이용해 현재 페이지 수 관리

- PostList.tsx 수정
  - PostList.tsx 내부에서 훅 사용시, PostListNav.tsx가 새로운 데이터 패칭을 트리거 했을 경우,
    PostList.tsx가 리렌더링 되지 않음
     - props을 통해 리스트 출력에 필요한 데이터를 내려받도록 변경

- PostListNav.tsx 수정
  - currentPageNumber, maxPageNumber를 props로 내려받도록 변경
  - currentPageNumber에 따라 Button을 비활성화 하기 위한 로직 작성 및 분리

- /board/post/index.tsx : 데이터 패칭의 주체로 변경

Rengod95 committed:
fix : remove unessential State

- useEffect로 state를 통해 또다른 state를 동기화 하는 코드를 수정했습니다.

- 일반 변수로 리팩토링 수행했습니다

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,46 @@
 
-export const GET_POST_QUERY_KEY = 'post';
-export const GET_POST_ENDPOINT = 'https://dummyjson.com/posts/';
+export type Post = { id: number; body: string; title: string };
+export type PostResponse = { posts: Post[]; total: number; limit: number; skip: number };
```
prettier 동작하는지 확인해주세요.

Rengod95 replied:
prettier 는 계속 작동하고 있습니다! 혹시 어떤 부분이 잘못 되었는지 알 수 있을까요??

imnotmoon replied:
줄바꿈이 될거라 생각했는데 안되고 있어서 여쭤봤습니다. 제 설정이 다른걸수도 있으니 한번 확인해볼게요. 일단은 넘어가주세요.

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,46 @@
 
-export const getPostAsync = async (offset: number = 0): Promise<any> => {
-  return await fetch(GET_POST_ENDPOINT + offset).then((res) => res.json());
+export const getPostAsync = async (pageParam: number = 1): Promise<PostResponse> => {
```
Async라는 수식어가 붙은 함수가 있으면 async하지 않은 함수도 있나? 하고 찾게 됩니다.
없다면 지워주세요. getPost 정도만으로도 충분합니다.

Rengod95 replied:
넵 반영해서 수정하겠습니다!

imnotmoon replied:
수정이 완료되면 코멘트로 commit hash를 남겨주세요. 그래야 수정여부를 확인하기가 쉬워요.

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,46 @@
+export const getPostAsync = async (pageParam: number = 1): Promise<PostResponse> => {
+  const skipCount = (pageParam - 1) * DEFAULT_POST_ITEM_SKIP;
+  const limitCount = pageParam * DEFAULT_POST_ITEM_LIMIT;
+  return await fetch(`${POSTS_ENDPOINT}?skip=${skipCount}&limit=${DEFAULT_POST_ITEM_LIMIT}`).then(
+    (res) => res.json()
+  );
```
query param을 template literal로 직접 만드는건 보안상 위험합니다.
아마 qs 모듈을 깔아놨을텐데 `qs.stringify()`의 사용법을 찾아서 대체해주세요.

Rengod95 replied:
아직 정확한 api가 나온 게 없어 신중하게 작성하지 않은 것 같습니다. qs 사용해서 수정해 보겠습니다!

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,46 @@
+    [GET_POST_QUERY_KEY, currentPageNumber],
+    () => getPostAsync(currentPageNumber)
   );
+  const maxPageNumber: number = data ? data.total / DEFAULT_POST_ITEM_LIMIT : 0;
```
data는 받았지만 total이 undfined일 수도 있습니다.
그럴 가능성은 매우 적긴 하지만 방어적으로 작성하는게 좋을것 같아요.

data?.total ? data.total / DEFAULT_POST_ITEM_LIMIT : 0;

Rengod95 replied:
넵! fetching data 를 조금 더 신중히 고려하는 습관을 길러보겠습니다

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,46 @@
-export const useInfinitePost = (id: number) => {
-  const getPostAsync = React.useCallback(() => getPostAsync, []);
+export const usePosts = () => {
+  const [currentPageNumber, setCurrentPageNumber] = useState<number>(1);
```
[fetchNextPage()](https://tanstack.com/query/v4/docs/react/reference/useInfiniteQuery#:~:text=with%20fetchPreviousPage.-,fetchNextPage,-%3A%20(options%3F%3A%20FetchNextPageOptions)%20%3D%3E%20Promise)를 쓰면 state 없이 다음 페이지 요청이 가능합니다.
리팩토링 해보시고 어려우면 놔두셔도 괜찮아요. 제가 수정 이어서 할게요.

Rengod95 replied:
도전해 보겠습니다!!

imnotmoon commented on src/components/Post/PostList/PostList.tsx:
```diff
@@ -1,42 +1,27 @@
 
 export type PostListProps = {
-  [k: string]: any;
+  data: PostResponse | undefined;
```
Response는 getXXX 함수같은 fetch단까지만 사용하고 멈춰야 합니다.

사실 컴포넌트가 prop으로 받아 사용하는 데이터 모델은 response를 그대로 사용하는게 가장 좋은데, 실제로는 그렇지 못할 경우가 많아요. 그래서 Response 타입을 만든거고 이건 fetch 함수에서 정리한 다음에 정제된 타입(뷰모델)을 쓰는게 맞습니다.

imnotmoon replied:
object를 그대로 컴포넌트에 넘겨주는것도 되도록 지양해주세요. 무차별적인 리렌더링이 발생합니다.
객체를 꼭 써야겠으면 spread 같은걸로 풀어헤친 다음에 내려주세요.
배열은 꼭 필요한 경우에는 허용합니다.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.css.ts:
```diff
@@ -1,21 +1,20 @@
   height: 80px;
 `;
-export const PostNavLayout = styled(FlexBoxCreator.rowStartFlexBox('div'))`
+export const NavBody = styled(FlexBoxCreator.rowStartFlexBox('div'))`
```
NavBody라고 굳이 쓸 필요가 없습니다.
어차피 className 해싱돼서 샌드박싱 됩니다.

Rengod95 replied:
아 NavBody라고 작성했던 이유는 혹여 PostListNav.tsx 의 Root컴포넌트 내부에 nav element를 제외한 다른 컴포넌트가 추가될 상황이 생길 수 있는걸 감안해서 접두사로 Nav를 붙였습니다.  의미가 없는 네이밍인가용?

imnotmoon replied:
PostListNav 컴포넌트에 nav element 말고 다른게 들어오면 그때부턴 PostListNav가 아닌겁니다.
애초에 고려할 필요가 없는 케이스예요.
NavBody 말고 Body면 충분합니다.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.hooks.ts:
```diff
@@ -1 +1,18 @@
-export const usePostListNav = () => {};
+import { useState, useEffect } from 'react';
+
+export const usePostListNav = (current: number, max: number) => {
```
이 컴포넌트는 필요 없어보여요.
inifinite scroll로 구현하라 하면 무용지물이 됩니다. 그냥 처음부터 안만드는게 마음 편해요.

Rengod95 replied:
일단 이것저것 만들어보고 싶은 제 의지랑 다르게 정확한 디자인이 없어서 그런것 같습니다 ㅠㅠ

imnotmoon replied:
네 의욕 충분히 이해합니다..
근데 나중에 디자인과 기획이 나오고 나서 보면 이미 만들어둔걸 수정하는 비용보다 새로 만드는 비용이 더 적게 들어요.
그러니 꼭 필요한거 말고는 작업을 안해두시는게 좋습니다.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,26 +1,49 @@
+type PostListNavProps = {
+  current: number;
+  max: number;
+  handler: {
```
함수를 props로 내려주는건 별로 좋은 생각이 아닙니다.

Rengod95 replied:
앗 저도 작성하면서 고민했던 부분입니다. PostList 에서 위 코드처럼 함수를 내려주는 게 아닌 다른 방식으로 PostListNav 가 다음 페이지 혹은 이전 페이지의 데이터 패칭을 트리거하게 하려면 어떻게 하는 방법이 최선인가요?

imnotmoon replied:
- 이전 / 다음 페이지 fetching을 위한 함수를 prop으로 내려주는 방법
  - useCallback으로 감싸진 onPreviousClick, onNextClick 함수를 내려주는게 좋습니다.
  - 똑같은 함수가 계속 생기는걸 막아줍니다. + 지금처럼 객체로 감싸서 내려주는건 최악입니다. 객체도 새로 생기고 함수도 새로 생겨요.
- prop으로 안받고 PostListNav에서 query를 호출해서 직접 fetchNextPage, fetchPreviousPage를 가져오는 방법
  - 재사용 관점에서 봤을때 PostListNav가 다른곳에서 재사용될 가능성이 있습니다.
  - 페이지네이션은 리스트의 책임입니다.
  - 그러니 PostListNav에서 들고있는건 적절하지 않은것 같습니다. 기각.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,26 +1,49 @@
-      </PostNavLayout>
-    </PostNavContainer>
+    <S.Root>
+      <S.NavBody>
```
NavBody -> Body

Rengod95 replied:
반영하겠습니다!

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,26 +1,49 @@
-    </PostNavContainer>
+    <S.Root>
+      <S.NavBody>
+        <S.ItemContainer>
```
container 굳이 필요 없어보여요.
필요하다면 styled-component 말고 일반 태그를 쓰고, 스타일을 먹여야 하면 selector로 접근해서 먹여주세요.

Rengod95 replied:
넵 이후 Pr에 수정 반영되도록 하겠습니다.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,26 +1,49 @@
+          </DefaultButton>
+        </S.ItemContainer>
+        <S.ItemContainer>
+          <S.ItemList>
```
ItemList -> List

imnotmoon replied:
List인데 자식 요소로 li가 없는게 어색해요. List면 ol 혹은 ul이어야 하고 그 자식으로 오는 요소는 리스트 요소일거라는게 일반적인 예상입니다.

Rengod95 replied:
감사합니다! 반영하여 수정하도록 하겠습니다!

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,26 +1,49 @@
+          </S.ItemList>
+        </S.ItemContainer>
+        <S.ItemContainer>
+          <DefaultButton
```
Default라는 prefix도 필요 없어보여요. 가독성을 해칠 수 잇으니 되도록 간단하게 써주세요.

Rengod95 replied:
넵!!

imnotmoon commented on src/components/commons/Button/button.type.ts:
```diff
@@ -17,7 +18,7 @@ export interface ButtonSize {
   height: number;
   padding?: number;
   margin?: number;
-  fontSize?: number;
+  fontSize?: string;
```
px 단위 말고 다른 단위도 받기 위해서 string으로 바뀐걸까요?

Rengod95 replied:
네 맞습니다! 현재 시점에서는 fontSize 뿐만 아니라, 프로젝트 내부의 Button, Icon, Flex 와 같은 스타일링 컴포넌트에서 공동으로 관리하는 CSS 타입에 대해 재차 고민해봐야 할 것 같습니다.

imnotmoon replied:
좋은 처리네요. 고민해보시고 깔끔하게 처리해서 PR 올려봐주세요 👍

imnotmoon commented on src/components/commons/Button/button.util.ts:
```diff
@@ -64,12 +62,11 @@ export const getButtonSize = (size: ButtonSize | ButtonSizePreset) => {
-    margin:${totalSize.margin}
-    padding:${totalSize.padding}
-    font-size:${totalSize.fontSize}
+    width: ${totalSize?.width + 'px'};
+    height: ${totalSize?.height + 'px'};
+    margin: ${totalSize?.margin + 'px'};
+    padding: ${totalSize?.padding + 'px'};
+    font-size: ${totalSize?.fontSize};
```
width, height, margin, padding, fontSize가 undefined로 들어오면 어떻게 동작하던가요?
확인은 못해봤는데 0으로 override 될 수도 있겠다 싶어서요.

Rengod95 replied:
size prop에  preset string 형태가 아닌 객체를 전달할 경우, width, height 는 필수 입력값이라, undefined가 들어가면 런타임 이전에 타입스크립트가 잡아줍니다. margin padding font-size의 경우 테스트 해본 결과 해당 css가 아예 안먹힙니다.

imnotmoon replied:
preset을 왜 두는지를 먼저 생각해보면 좋을것 같아요.

preset이 필요한건 공유받은 디자인을 싹 봤을때 버튼 사이즈가 정규화된 경우에만 유효해요.
제일 좋은건 정규화돼서 small, medium, large같은 사이즈로 표현된 버튼 외에는 쓸 일이 없는거에요. 근데 이건 디자이너 입장에서 불가능에 가까워요.
그 다음으로 일반적인건 어느정도 사이즈 정규화가 되어있지만 일부 outlier가 있는겁니다. 이런 경우는 인님 방법대로 직접 스타일을 부여해야만 해요. 근데 보통은 버튼 사이즈를 잡을때 width, height를 직접 주지 않고 padding과 line-height 정도만 사용합니다. 반응형 처리할때 불리하기 때문이예요.
그리고 컴포넌트에 직접 width, height를 적어주면 미관상 좋지도 않아요. 스크립트의 책임과 스타일시트의 책임에 대해서 한번 더 생각해봐주세요.

imnotmoon commented on src/components/commons/Header/Header.css.ts:
```diff
@@ -40,7 +40,7 @@ export const NavListItem: StyledComponent<FlexListItemProps> = styled(FlexListIt
 `;
 
-export const StyledLink = styled(FlexBoxCreator.createDefaultFlexBox(Link))`
+export const RouterLink = styled(FlexBoxCreator.createDefaultFlexBox(Link))`
```
얘는 selector로 스타일 먹여도 충분할것 같네요.
물론 아니어도 괜찮습니다. 이 코멘트는 인님 생각에 따라 반영 여부 결정해주세요.

imnotmoon commented on src/components/commons/Header/Header.tsx:
```diff
@@ -14,16 +14,16 @@ export const NAV_ITEM_DATAS: Array<NavItemData> = [
-    <Root>
-      <Nav>
-        <NavItemList>
+    <S.Root>
+      <S.Body>
+        <S.HeaderItemList>
```
보기 좋아졌어요.

Rengod95 committed:
fix: PostListHook.ts Infinite로 변경

- usePost() -> useInfinitePost로 이름 변경
- useQuery 기반의 페이지네이션에서useInfiniteQuery 기반으로 변경
- usePostListNav 훅 삭제 및 페칭 트리거 동작을 전부 Infinite Query가 제공해주는 함수를 사용하도록 변경

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,44 @@
+    limit: DEFAULT_POST_ITEM_LIMIT,
+  });
+
+  const data = await fetch(`${POSTS_ENDPOINT}?${queryString}`).then((res) => res.json());
```
[request](https://github.com/Rengod95/Dynamic_Roadmap/blob/83cf9e9f160e017ffda2d76b6da1cd8d127f2abc/src/utils/fetch.ts#L4) 써주시면 더 좋을것 같습니다.

imnotmoon commented on src/components/Post/PostList/PostList.hooks.ts:
```diff
@@ -1,32 +1,44 @@
-    isLoading,
-    error,
-  };
+export const getInfinitePost = async (pageParam: number): Promise<PostResponse> => {
```
pageParam을 받는다 === infinite라는게 성립하지는 않는것 같아요.
pageParam을 optional로 처리하고 undefined가 아니면 해당 페이지를 가져오고, 아니면 그냥 첫 페이지를 가져오게 처리하면 어떨까요?
너무 똑똑해서 유지보수가 어렵지도 않을거고 infinite를 뗄 수도 있을것 같아요.

imnotmoon commented on src/components/Post/PostList/PostList.tsx:
```diff
@@ -1,43 +1,32 @@
-        {isLoading && <p>Loading user data...</p>}
-        {isError && <p>Error fetching user data: {error?.message && '에러'}</p>}
-        {data && <div>data</div>}
+        {query.isLoading ? (
+          <div>is loading...</div>
```
[Suspense](https://beta.reactjs.org/reference/react/Suspense)

imnotmoon commented on src/components/Post/PostList/PostList.tsx:
```diff
@@ -1,43 +1,32 @@
-        {data && <div>data</div>}
+        {query.isLoading ? (
+          <div>is loading...</div>
+        ) : query.isError ? (
+          <div>error was caught!</div>
```
[ErrorBoundary](https://reactjs.org/docs/error-boundaries.html)

imnotmoon commented on src/components/Post/PostList/PostList.tsx:
```diff
@@ -1,43 +1,32 @@
+            return page.posts.map((page) => {
+              return (
+                <div key={page.id}>
+                  <h2>{page.id}</h2>
+                  <h3>{page.title}</h3>
```
h1 - 애플리케이션 / 서비스명
h2 - 서비스명
h3 - 페이지명
h4 - 컴포넌트 타이틀
h5 - 컴포넌트 in 컴포넌트의 타이틀

이정도로 정리하기로 해요. h2와 h3는 너무 높은것 같습니다.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,27 +1,37 @@
+          size={'medium'}
+          variant={'default'}
+          disabled={!query.hasPreviousPage || query.isFetchingPreviousPage}
+          onClick={() => query.fetchPreviousPage()}>
```
일반적인 action과 event handler를 구분할 필요가 있어요.
여기서 action은 query.fetchPreviousPage()이고 event handler는 onClick은 () => query.fetchPreviousPage()입니다.

근데 둘 사이의 관계를 보면 이렇게 돼요. 
event handler has action (O)
event handler is action (X)

onClick 함수를 만들어서 query.fetchPreviousPage를 넣어주고, onClick은 만들어진 onClick을 넘기는게 좋겠어요.

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,27 +1,37 @@
+        </DefaultButton>
+        <span>
+          {query.data?.pageParams.map((page) => (
+            <span key={page as number}>{page as ReactNode}</span>
```
의심스러운 type assertion이 들어가있네요. 꼭 필요한가요?

imnotmoon commented on src/components/Post/PostListNav/PostListNav.tsx:
```diff
@@ -1,27 +1,37 @@
+        <DefaultButton
+          size={'medium'}
+          variant={'default'}
+          disabled={!query.hasNextPage || query.isFetchingNextPage}
```
query 객체에 속한 data, hasPreviousPage, isFetchingPreviousPage, hasNextPage, isFEtchingNextPage는 구조분해 해서 접근해도 되는데 굳이 query를 거쳐서 접근하고 있어요. 미관상 좋지 않으니 안전하게 구조분해 해서 사용해주세요.

imnotmoon commented on src/components/commons/Button/Button.css.ts:
```diff
@@ -1,15 +1,16 @@
-  ${({ size, variant }: ButtonProps) => {
-    const colorStyle = getButtonColor(variant);
-    const sizeStyle = getButtonSize(size);
+export const Button = styled(FlexBoxCreator.createDefaultFlexBox('button'), {
```
button은 왜 그냥 button이 아니고 얘도 flexbox여야만 하나요?

imnotmoon replied:
지금 열려있는 PR이 전부 마무리되면 불필요한 Flex 다 걷어내는 작업 진행해주세요.
flex는 레이아웃을 잡을때 말고는 사용 불허하겠습니다.

imnotmoon commented on src/components/commons/Button/button.util.ts:
```diff
@@ -1,11 +1,11 @@
-  small: { width: 100, height: 50 },
-  medium: { width: 200, height: 60 },
-  large: { width: 300, height: 80 },
+  none: { width: 0, height: 0, margin: 0, padding: 0, fontSize: '14px' },
+  small: { width: 100, height: 50, margin: 0, padding: 0, fontSize: '14px' },
+  medium: { width: 200, height: 60, margin: 0, padding: 0, fontSize: '14px' },
+  large: { width: 300, height: 80, margin: 0, padding: 0, fontSize: '14px' },
```
font-size는 도저히 지금 필요 없는 속성인것 같아요.

imnotmoon commented on src/pages/board/post/index.tsx:
```diff
@@ -1,21 +1,36 @@
+
+  return {
+    props: {
+      dehydratedState: JSON.parse(JSON.stringify(dehydrate(queryClient))),
```
JSON.stringify -> JSON.parse 과정이 왜 필요한가요?
